File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.max_dist = 0
    
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        self.traverse(root, start)
        return self.max_dist
    
    def traverse(self, root: Optional[TreeNode], start: int) -> int:
        depth = 0
        
        # base case: if it is a leaf
        if not root:
            return depth
        
        left_depth = self.traverse(root.left, start)
        right_depth = self.traverse(root.right, start)
        
        # when the current node reaches to start point
        if root.val == start:
            self.max_dist = max(left_depth, right_depth)
            depth = -1
            
        elif left_depth >= 0 and right_depth >= 0:
            depth = max(left_depth, right_depth) + 1
            
        else:
            dist = abs(left_depth) + abs(right_depth)
            self.max_dist = max(self.max_dist, dist)
            depth = min(left_depth, right_depth) - 1
        
        return depth

File: 2385-amount-of-time-for-binary-tree-to-be-infected/2385-amount-of-time-for-binary-tree-to-be-infected.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int amountOfTime(TreeNode* root, int start) {
        queue<TreeNode*> q{{root}};
        unordered_map<int, vector<int>> adj;
        
        while (!q.empty()){
            TreeNode* node = q.front();
            q.pop();
            
            if (node -> left){
                adj[node -> val].push_back(node -> left -> val);
                adj[node -> left -> val].push_back(node -> val);
                q.push(node -> left);
            }
            
            if (node -> right){
                adj[node -> val].push_back(node -> right -> val);
                adj[node -> right -> val].push_back(node -> val);
                q.push(node -> right);
            }
        }
        
        int ans = 0;
        vector<bool> visited(100'001);
        queue<pair<int, int>> graphQ;
        graphQ.push(make_pair(start, 0));
        while(!graphQ.empty()){
            auto [node, dist] = graphQ.front();
            graphQ.pop();
            
            if (!visited[node]){
                ans = max(ans, dist);
                visited[node] = true;
                for (int neighbor: adj[node]){
                    graphQ.push({neighbor, dist + 1});
                }
            }
        }
        
        return ans;
    }
};